runtime/src/plugin.rs

```diff
@@ -251,44 +251,75 @@ impl Plugin {
         self.memory.store.data().wasi.is_some()
     }
 
-    pub fn initialize_runtime(&mut self) -> Result<(), Error> {
+    fn detect_runtime(&mut self) -> Option<Runtime> {
+        // Check for Haskell runtime initialization functions
         // Initialize Haskell runtime if `hs_init` and `hs_exit` are present,
         // by calling the `hs_init` export
         if let Some(init) = self.get_func("hs_init") {
-            if self.get_func("hs_exit").is_some() {
-                // as a final check, ensure that the `hs_init` signature meets the exepectation
+            if let Some(cleanup) = self.get_func("hs_exit") {
                 if init.typed::<(i32, i32), (), _>(&self.memory.store).is_err() {
-                    return Ok(());
+                    trace!("hs_init found, but has the incorrect type");
```

Comment:
maybe we could instead serialize and log the function type here vs. saying "incorrect" -- if it's in there it could be correct to someone, just not for Haskell runtime detection.